File: Data Structures & Algorithms/reorder-linked-list/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        s=f=head
        while f and f.next:
            if not f.next:
                break
            s=s.next
            f=f.next.next
        temp=s.next
        s.next=None
        s=temp

        curr=s
        pre=None
        while curr:
            temp=curr.next
            curr.next=pre
            pre=curr
            curr=temp
        
        l1=head
        l2=pre

        while l1 and l2:

            t1=l1.next
            l1.next=l2

            t2=l2.next
            l2.next=t1

            l1=t1
            l2=t2
        

        
            

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-3.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if(lists.length==0) return null;
        
        ListNode merge=lists[0];
        for(int i=1;i<lists.length;i++){
            merge=merger(merge,lists[i]);
        }
        return merge;
    }
    public ListNode merger(ListNode l1,ListNode l2){
        ListNode dummy=new ListNode(0);
        ListNode temp=dummy;
        while(l1!=null && l2!=null){
            if(l1.val<=l2.val){
                temp.next=l1;
                l1=l1.next;
            }
            else {
                temp.next=l2;
                l2=l2.next;
            }
            temp=temp.next;
        }
        temp.next=(l1==null)?l2:l1;
        return dummy.next;
    }
}
